src/mtype.d

```diff
@@ -6199,10 +6204,19 @@ extern (C++) final class TypeFunction : TypeNext
                     fparam.storageClass |= STCreturn;
                 }
 
-                if (fparam.storageClass & STCreturn && !(fparam.storageClass & (STCref | STCout)))
+                if (fparam.storageClass & STCreturn)
                 {
-                    error(loc, "'return' can only be used with 'ref' or 'out'");
-                    errors = true;
+                    if (!(fparam.storageClass & (STCref | STCout)))
+                    {
+                        error(loc, "'return' can only be used with 'ref' or 'out'");
+                        errors = true;
+                    }
+                    else if (!tf.isref && tf.next && !tf.next.hasPointers())
+                    {
+                        error(loc, "parameter %s is 'return' but function does not return any indirections",
```

Comment:
> may also need a test where there are two return parameters and only one of them is returned unconditionally.

I don't agree. It's about the interface, not the implementation. One should be able to change the implementation and keep the same interface, such as with virtual overrides.